benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,281 @@
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/ginkgo.hpp>
+
+
+#include <chrono>
+#include <memory>
+
+
+#include <gflags/gflags.h>
+
+
+#ifdef HAS_CUDA
+
+
+#include <cuda.h>
+#include <cuda_runtime.h>
+
+
+#include "cuda/base/device_guard.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+#ifdef HAS_HIP
+
+
+#include <hip/hip_runtime.h>
+
+
+#include "hip/base/device_guard.hip.hpp"
+
+
+#endif  // HAS_CUDA
+
+
+DEFINE_bool(gpu_timer, false,
+            "use gpu timer based on event. It is valid only when "
+            "executor is cuda or hip");
+
+
+class Timer {
+public:
+    void tic()
+    {
+        assert(tic_called_ == false);
+        this->tic_impl();
+        tic_called_ = true;
+    }
+
+    std::size_t toc()
+    {
+        assert(tic_called_ == true);
+        auto ns = this->toc_impl();
+        tic_called_ = false;
+        this->add_record(ns);
+        return ns;
+    }
+
+    std::size_t get_total_time() { return total_duration_ns_; }
+
+    std::size_t get_tictoc_num() { return duration_ns_.size(); }
+
+    double get_average_time()
+    {
+        return static_cast<double>(this->get_total_time()) /
+               this->get_tictoc_num();
+    }
+
+    void clear()
+    {
+        duration_ns_.clear();
+        tic_called_ = false;
+        total_duration_ns_ = 0;
+    }
+
+    Timer() : tic_called_(false), total_duration_ns_(0) {}
+
+protected:
+    void add_record(std::size_t ns)
+    {
+        // add the result;
+        duration_ns_.emplace_back(ns);
+        total_duration_ns_ += ns;
+    }
+
+    virtual void tic_impl() = 0;
+
+    virtual std::size_t toc_impl() = 0;
+
+private:
+    std::vector<std::size_t> duration_ns_;
+    bool tic_called_;
+    std::size_t total_duration_ns_;
```

Comment:
I would prefer if we store this as `std::int64_t` (signed).
I am fine with storing this as nanoseconds because we don't lose precision when converting from `float` (or even `double`) to this, assuming we never measure something close to the nanosecond range (and if we would, we need a new implementation anyway because a virtual function call is too slow for that).
However, I would prefer a signed type for the same reason we want to use signed as the new `size_type` in Ginkgo 2.0:
- for 64 bit, we will never use the extra bit (like this, we can already measure up to 300 years)
- we prevent the many warnings for casting from unsigned to signed
Also, as @Slaedr already pointed out, `std::chrono` already uses a signed type (`std::int64_t`) for duration, so we prevent a casting warning here as well.

Reply:
Technically `float` can go up to 10^49 or something which would overflow even `size_t`, and the GPU timers return `float`, I would consider just using `double`. I agree that in practice, we can probably also just use `int64_t`. :)

Reply:
Yes, `float` can fit that, but we lose a lot of resolution in the process with float, which we won't with `int64`. Using plain `double` would obviously also work, and I would be fine with that as well.